File: backend/controllers/notifications.js
const PushToken = require('../models/PushToken');
const logger = require('../config/logger');

exports.registerToken = async (req, res, next) => {
  try {
    const { token, platform } = req.body;

    if (!token || !platform) {
      return res.status(400).json({ success: false, error: 'Token and platform are required' });
    }

    if (!['ios', 'android', 'web'].includes(platform)) {
      return res.status(400).json({ success: false, error: 'Invalid platform' });
    }

    await PushToken.findOneAndUpdate(
      { token },
      { userId: req.user.id, token, platform },
      { upsert: true, new: true }
    );

    logger.info({ userId: req.user.id, platform }, 'Push token registered');
    res.status(200).json({ success: true, data: 'Token registered' });
  } catch (err) {
    next(err);
  }
};

exports.unregisterToken = async (req, res, next) => {
  try {
    const { token } = req.body;

    if (!token) {
      return res.status(400).json({ success: false, error: 'Token is required' });
    }

    await PushToken.findOneAndDelete({ token, userId: req.user.id });

    logger.info({ userId: req.user.id }, 'Push token unregistered');
    res.status(200).json({ success: true, data: 'Token unregistered' });
  } catch (err) {
    next(err);
  }
};

File: backend/app/routers/notifications.py
import logging

from fastapi import APIRouter, Depends, HTTPException

from app.db import get_db
from app.dependencies import get_current_user
from app.schemas import PushTokenBody, UnregisterTokenBody

router = APIRouter(prefix="/api/notifications", tags=["notifications"])
logger = logging.getLogger("bloodconnect.notifications")


@router.post("/register-token")
def register_token(body: PushTokenBody, user=Depends(get_current_user)):
    db = get_db()
    db.pushtokens.update_one(
        {"token": body.token},
        {"$set": {"userId": user["_id"], "token": body.token, "platform": body.platform}},
        upsert=True,
    )
    logger.info("Push token registered for user %s on %s", user["_id"], body.platform)
    return {"success": True, "data": "Token registered"}


@router.delete("/unregister-token")
def unregister_token(body: UnregisterTokenBody, user=Depends(get_current_user)):
    if not body.token:
        raise HTTPException(status_code=400, detail={"success": False, "error": "Token is required"})
    get_db().pushtokens.delete_one({"token": body.token, "userId": user["_id"]})
    logger.info("Push token unregistered for user %s", user["_id"])
    return {"success": True, "data": "Token unregistered"}
